qd_proto/dissectors/qd.lua

```diff
@@ -0,0 +1,169 @@
+local function check_input_buf_len(buf, off)
+    return ((off < 0) or (off >= buf:len())) and false or true
+end
+
+-- Check if input buffer was sliced/cut-off.
+-- @param buf Input buffer.
+-- @param off Offset in input buffer.
+-- @return true  - if input buffer whole;
+--         false - if not.
+local function check_input_buf_cut_off(buf, off)
+    if ((buf:len()) - off ~= buf:reported_length_remaining(off)) then
+        return false
+    end
+    return true
+end
+
+-- Check compact length contained in buf.
+-- @param buf Input buffer.
+-- @param off Offset to compact length.
+-- @return true  - if compact length may fit in buffer;
+--         false - if not.
+local function check_compact_len(buf, off)
+    local n = buf(off, 1):uint()
+    local compact_len = utils.get_compact_len(n)
+    local buf_len = buf:len() - off
+    if (compact_len > buf_len) then
+        return false
+    else
+        return true
+    end
+end
+
+-- Check message length contained in buf.
+-- @param buf Input buffer.
+-- @param off Offset to message length.
+-- @return true  - if message length success read;
+--         false - if not.
+local function check_msg_len(buf, off)
+    local msg_len = utils.read_compact_int(buf, off)
+    if (msg_len == nil or msg_len < 0) then
+        return false
+    else
+        return true
+    end
+end
+
+-- Get message type.
+-- @param buf Input buffer.
+-- @param off Offset to message type.
+-- @return Message type.
+local function get_msg_type(buf, off)
+    local type = {}
+    -- Message type num.
+    type.val_uint = buf(off, 1):uint()
+    -- String representation message type.
+    type.val_str = utils.enum_val_to_str(fields.message_type, type.val_uint)
+    return type
+end
+
+-- Check receive message.
+-- @param buf Input buffer.
+-- @param off Offset in input buffer.
+-- @return  0  - this not QD message;
+--         <0  - size input buffer not enough for assemble
+--               full QD message (negative num bytes, num which are missing);
+--         >0, - full size QD message.
+function qd.check_msg(buf, off)
+    local result = check_input_buf_len(buf, off)
+    if (result == false) then
+        dbg.warn(dbg.file(), dbg.line(), "Bad input buffer length.")
+        return 0
+    end
+
+    local result = check_input_buf_cut_off(buf, off)
+    if (result == false) then
+        dbg.warn(dbg.file(), dbg.line(),
+                 "Captured packet was shorter than original, can't reassemble.")
+        return 0
+    end
+
+    local result = check_compact_len(buf, off)
+    if (result == false) then
+        dbg.info(dbg.file(), dbg.line(),
+                 "Not enough buffer length for get compact_len.")
+        return -DESEGMENT_ONE_MORE_SEGMENT
+    end
+
+    local result = check_msg_len(buf, off)
+    if (result == false) then
+        dbg.warn(dbg.file(), dbg.line(), "Bad message length.")
+        return 0
+    end
+
+    local msg_len, msg_len_size = utils.read_compact_int(buf, off)
```

Comment:
There is no handling of the case when `read_compact_int` returns `nil` (`compact_len > buf:len() - off`).